lapack/testlapack/dorgr2.go

```diff
@@ -0,0 +1,67 @@
+// Copyright ©2021 The Gonum Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package testlapack
+
+import (
+	"testing"
+
+	"golang.org/x/exp/rand"
+
+	"gonum.org/v1/gonum/floats/scalar"
+)
+
+type Dorgr2er interface {
+	Dgerqfer
+	Dorgr2(m, n, k int, a []float64, lda int, tau []float64, work []float64)
+}
+
+func Dorgr2Test(t *testing.T, impl Dorgr2er) {
+	rnd := rand.New(rand.NewSource(1))
+	for ti, test := range []struct {
+		// Dorgr2 lim: n >= m >= k >= 0
+		m, n, lda int
+	}{
+		{2, 2, 0}, // constructQK only creates square matrices.
+		{3, 3, 0},
+		{30, 30, 0},
+	} {
+		m := test.m
+		n := test.n
+		lda := test.lda
+		if lda == 0 {
+			lda = n
+		}
+		a := randomGeneral(m, n, lda, rnd)
+
+		// Compute the QR decomposition of A.
+		k := min(m, n)
+		tau := make([]float64, k)
+		work := make([]float64, 1)
+		impl.Dgerqf(m, n, a.Data, a.Stride, tau, work, -1)
+		work = make([]float64, int(work[0]))
+		impl.Dgerqf(m, n, a.Data, a.Stride, tau, work, len(work))
+
+		// Compute the matrix Q explicitly using the first k elementary reflectors.
+		q := constructQK("RQ", m, n, k, a.Data, a.Stride, tau)
+
+		// Compute the matrix Q using Dorg2r.
+		impl.Dorgr2(m, n, k, a.Data, a.Stride, tau, work)
+
+		// Check that the first n columns of both results match.
+		same := true
```

Comment:
In our convention we use "same" to mean "both inputs are equal or both are NaN". But you should rather compute residuals so that the decision pass/fail is done on a single indicator which we can log and see how far the result is from the expectation.